### PR TITLE
Bootloader: OSX: Respect Info.plist options

### PR DESCRIPTION
**Pull request by BoboTiG**
A possible fix for #1917.

## What is the fix/workaround?

As proposed by @codewarrior0 in his [comment](https://github.com/pyinstaller/pyinstaller/pull/1206#issuecomment-165976396), I added a bootloader runtime option `pyi-osx-background` that triggers the right transform type in `pyi_launch_initialize()`.

This is the simplest option as it does not require to add new bootloaders nor a new argument to handle this specific case. The behavior is restored as it should have been: when the user sets `LSUIElement` in the Info.plist dict, PyInstaller handles it.

## Current status

The patch is good but does not work automatically because the TOC option is not yet forwarded to the bootloader.

In `$project/build/BUNDLE-00.toc`, I have the line `('pyi-osx-background', '', 'OPTION')` setted, which is what we need.
But as this kind of TOC line is traited before (when building the PKG), it is not possible to get it via `pyi_arch_get_option(status, "pyi-osx-background")`, steps are:
```
289 INFO: PyInstaller: 3.4.dev0+g8c071568.mod
290 INFO: Python: 3.6.2
297 INFO: Platform: Darwin-16.7.0-x86_64-i386-64bit
299 INFO: UPX is not available.
301 INFO: Extending PYTHONPATH with paths
['.../pyinstaller']
301 INFO: checking Analysis
301 INFO: Building Analysis because Analysis-00.toc is non existent
302 INFO: Initializing module dependency graph...
303 INFO: Initializing module graph hooks...
305 INFO: Analyzing base_library.zip ...
4519 INFO: running Analysis Analysis-00.toc
4533 INFO: Caching module hooks...
4538 INFO: Analyzing systray.py
4552 INFO: Loading module hooks...
4553 INFO: Loading module hook "hook-encodings.py"...
4627 INFO: Loading module hook "hook-pydoc.py"...
4627 INFO: Loading module hook "hook-PyQt5.py"...
4657 INFO: Loading module hook "hook-PyQt5.QtCore.py"...
4749 INFO: Loading module hook "hook-PyQt5.QtGui.py"...
4784 INFO: Loading module hook "hook-PyQt5.QtWidgets.py"...
4819 INFO: Loading module hook "hook-xml.py"...
5172 INFO: Looking for ctypes DLLs
5172 INFO: Analyzing run-time hooks ...
5177 INFO: Including run-time hook 'pyi_rth_qt5.py'
5189 INFO: Looking for dynamic libraries
5451 INFO: Looking for eggs
5451 INFO: Using Python library venv-pyinstaller/bin/../.Python
5457 INFO: Warnings written to systray/build/systray/warn-systray.txt
5495 INFO: Graph cross-reference written to systray/build/systray/xref-systray.html
5508 INFO: checking PYZ
5508 INFO: Building PYZ because PYZ-00.toc is non existent
5509 INFO: Building PYZ (ZlibArchive) systray/build/systray/PYZ-00.pyz
5887 INFO: Building PYZ (ZlibArchive) systray/build/systray/PYZ-00.pyz completed successfully.
5894 INFO: checking PKG
5894 INFO: Building PKG because PKG-00.toc is non existent
5894 INFO: Building PKG (CArchive) PKG-00.pkg
5904 INFO: Building PKG (CArchive) PKG-00.pkg completed successfully.
5905 INFO: Bootloader pyinstaller/PyInstaller/bootloader/Darwin-64bit/runw_d
5905 INFO: checking EXE
5905 INFO: Building EXE because EXE-00.toc is non existent
5905 INFO: Building EXE from EXE-00.toc
5905 INFO: Appending archive to EXE systray/build/systray/systray
5907 INFO: Fixing EXE for code signing systray/build/systray/systray
5915 INFO: Building EXE from EXE-00.toc completed successfully.
5917 INFO: checking COLLECT
5917 INFO: Building COLLECT because COLLECT-00.toc is non existent
5917 INFO: Building COLLECT COLLECT-00.toc
7202 INFO: Building COLLECT COLLECT-00.toc completed successfully.
7209 INFO: This is an agent app (LSUIElement is set to True)
7209 INFO: checking BUNDLE
7209 INFO: Building BUNDLE because BUNDLE-00.toc is non existent
7209 INFO: Building BUNDLE BUNDLE-00.toc
8430 INFO: moving BUNDLE data files to Resource directory
```
As we can see, the line `7209 INFO: This is an agent app (LSUIElement is set to True)` appears too late in the process.

Any clue to make it easy forwarded? Perhaps I misunderstand something?

### CODE REVIEW

File: PyInstaller/building/osx.py [hunk 1]
```diff
@@ -163,6 +168,8 @@ def assemble(self):
         links = []
         toc = add_suffix_to_extensions(self.toc)
         for inm, fnm, typ in toc:
+            if typ == 'OPTION':
+                continue
```
Comment (htgoebel) [inline PyInstaller/building/osx.py]: This is an unrelated change and shall go into a commit of it's own.

File: PyInstaller/building/osx.py [hunk 2]
```diff
@@ -53,7 +53,12 @@ def __init__(self, *args, **kws):
             # Fallback to appname.
             self.bundle_identifier = self.appname
 
-        self.info_plist = kws.get('info_plist', None)
+        self.info_plist = kws.get('info_plist', [])
```
Comment (htgoebel) [inline PyInstaller/building/osx.py]: If `info_plist` is not given as an argument, `self.info_plist` will become a list, and the next line will fail:
```
>>> [].get('a')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'list' object has no attribute 'get'
```

Comment (BoboTiG) [reply PyInstaller/building/osx.py]: Yeah, stupid mistake. 👍

File: PyInstaller/building/osx.py (hunk 1 above)
Comment (BoboTiG) [reply PyInstaller/building/osx.py]: I had to add this to prevent such exception:
```python
8090 INFO: Building BUNDLE BUNDLE-00.toc
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "pyinstaller/pyinstaller.py", line 15, in <module>
    run()
  File "pyinstaller/PyInstaller/__main__.py", line 109, in run
    run_build(pyi_config, spec_file, **vars(args))
  File "pyinstaller/PyInstaller/__main__.py", line 61, in run_build
    PyInstaller.building.build_main.main(pyi_config, spec_file, **kwargs)
  File "pyinstaller/PyInstaller/building/build_main.py", line 838, in main
    build(specfile, kw.get('distpath'), kw.get('workpath'), kw.get('clean_build'))
  File "pyinstaller/PyInstaller/building/build_main.py", line 784, in build
    exec(text, spec_namespace)
  File "<string>", line 45, in <module>
  File "pyinstaller/PyInstaller/building/osx.py", line 91, in __init__
    self.__postinit__()
  File "pyinstaller/PyInstaller/building/datastruct.py", line 158, in __postinit__
    self.assemble()
  File "pyinstaller/PyInstaller/building/osx.py", line 182, in assemble
    shutil.copy(fnm, tofnm)
  File "/lib/python3.6/shutil.py", line 241, in copy
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/lib/python3.6/shutil.py", line 120, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] No such file or directory: ''
```
But I will remove it because it will become useless after the next round.

**Comment by BoboTiG** [comment]
I updated the PR, it is working now.

Let me know your thoughts, do not hesitate to ask for minor changes (even in logged sentences). When it will be OK, I will add documentation and test cases.

**Comment by BoboTiG** [comment]
Finally I found a way simpler fix. By not setting ourselves `TransformProcessType()`, the OS handles it automatically.
I just tested several scenarii and all worked well:
- Without Info.plist: display normally with icon in the Dock.
- With Info.plist `LSUIElement=True`: display normally with no icon in the Dock.
- With Info.plist `LSBackgroundOnly=True`: no display nor icon in the Dock.

I think we have the desired behavior. Same results with `console=True`, Info.plist options are taken into account.

I am open to any advice/review to ensure I did not break something. Or else, I broke the bad behavior, but I want to be sure this is what we want.

**Comment by htgoebel** [comment]
Great, looks like you found the root cause. :-)

I wonder if `console=True` should set `LSUIElement` or `LSBackgroundOnly` in the `info_plist`? Or should we leave this to the developer? If we leave this to the developer, shall we add a description to the manual?

**Comment by BoboTiG** [comment]
I think it should automatically set `LSBackgroundOnly` (as stated in the [documentation](https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/LaunchServicesKeys.html#//apple_ref/doc/uid/20001431-109268), it seems the rght thing).

**Comment by BoboTiG** [comment]
Done in a separate commit. Do you prefer all in one commit?

Also, what about updated bootloaders? Should I provide them in a new commit?

File: PyInstaller/building/api.py [hunk 3]
```diff
@@ -354,6 +354,9 @@ def __init__(self, *args, **kwargs):
         else:
             self.upx = False
 
+        # Save the console value for later use on macOS
+        CONF['console'] = self.console
```
Comment (htgoebel) [inline PyInstaller/building/api.py]: No need for this. The `console` can be picked up in `EXE` at https://github.com/pyinstaller/pyinstaller/pull/3566/files#diff-c335c5207e1c5d707f757c131b913b62L62. And when doing this, `console` can be per EXE, while when storing the value in `CONFIG` it's global for all executable.

**Comment by htgoebel** [comment]
I would prefer to have this two commits: One for the Bootloader change and one for "Automatically set Info.plist LSBackgroundOnly=True when consol…".

Please do not build bootloaders, we have the policy that only bootloaders compiled by core develeopers are commited. And we need to recompile all of then soon anyway due to the upcoming Python 3.7 release.

File: PyInstaller/building/api.py (hunk 3 above)
Comment (BoboTiG) [reply PyInstaller/building/api.py]: I would prefer too, but condition `if isinstance(arg, EXE):` is never taken.

Comment (htgoebel) [reply PyInstaller/building/api.py]: IC. According to the templates (PyInstaller/building/templates.py), only for a one-file BUNDLE, the first argument is an EXE. For the one-dir Bundle, it is a COLLECT.

Two possible solutions:
1) In COLLECT there is already a loop scanning for EXE. Use this to set COLLECT's `.console` for the collection. This has the drawback, that we hide "a bundle can contain only one exe" in an unrelated area of the code.
2) In BUNDLE, if it is a COLLECT, scan the COLLECT's `.toc` for an exe and fetch the info from there. I just discover that [there is already a loop searching for Exe](https://github.com/BoboTiG/pyinstaller/blob/v3.3.1/PyInstaller/building/osx.py#L78). This loop takes into account the first exe only, thus I would do the same for `.console`.

Comment (BoboTiG) [reply PyInstaller/building/api.py]: The option 2 is not doable (as of now) because data in `toc` is just strings, for instance something like:
```python
('systray', '.../systray/build/systray/systray', 'EXECUTABLE')
```
If we have had an attribute pointing to the `EXE` object it would have been easy. But I do not think it is a good thing to do that.

I updated the PR to remove the use of `CONF` using option 1.

**Comment by BoboTiG** [comment]
Do you need changes on the PR?

**Comment by micahflee** [comment]
When this PR is merged, will it totally fix #1917? I'm looking forward to it!

**Comment by BoboTiG** [comment]
I should do the trick. Do you think you can check before any merging?

**Comment by micahflee** [comment]
Yes I'll give it a shot.

**Comment by micahflee** [comment]
I just tested myself, and it appears to work great! I also build my own bootloader, not sure if that was necessary or not. I can't wait for the next PyInstaller release.

**Comment by BoboTiG** [comment]
Great!

Yes, it is required to rebuild bootloaders in order to make the patch working.

File: PyInstaller/building/osx.py [hunk 4]
```diff
@@ -137,20 +140,16 @@ def assemble(self):
...
-
                            }
 
         # Merge info_plist settings from spec file
         if isinstance(self.info_plist, dict) and self.info_plist:
             info_plist_dict.update(self.info_plist)
 
+        # Setting EXE console=True implicitely sets LSBackgroundOnly to True
+        if self.console:
+            info_plist_dict['LSBackgroundOnly'] = True
```
Comment (htgoebel) [inline PyInstaller/building/osx.py]: If we move this in front of "Merge info_plist settings from spec file", this would allow users to overwrite the `LSBackgroundOnly` if needed for some reason. What do you think?

Comment (BoboTiG) [reply PyInstaller/building/osx.py]: Nice catch, change applied.

**Comment by htgoebel** [comment]
Thanks you very much for this nice pull-request and for finding this simple solution!

**Comment by BoboTiG** [comment]
And thanks for your patience and supervision.
Pleasure to contribute, I will keep going.
